core/src/stdlib/pyscript/events.py

```diff
@@ -0,0 +1,165 @@
+        """
+        for listener in self._listeners:
+            if is_awaitable(listener):
+                # Use create task to avoid making this an async function.
+                asyncio.create_task(listener(result))
+            else:
+                listener(result)
+
+    def add_listener(self, listener):
+        """
+        Add a callable/awaitable to listen to when this event is triggered.
+        """
+        if is_awaitable(listener) or callable(listener):
+            if listener not in self._listeners:
+                self._listeners.append(listener)
+        else:
+            raise ValueError("Listener must be callable or awaitable.")
+
+    def remove_listener(self, *args):
+        """
+        Clear the specified handler functions in *args. If no handlers
+        provided, clear all handlers.
+        """
+        if args:
+            for listener in args:
+                self._listeners.remove(listener)
+        else:
+            self._listeners = []
+
+
+def when(target, *args, **kwargs):
+    """
+    Add an event listener to the target element(s) for the specified event type.
+
+    The target can be a string representing the event type, or an Event object.
+    If the target is an Event object, the event listener will be added to that
+    object. If the target is a string, the event listener will be added to the
+    element(s) that match the (second) selector argument.
+
+    If a (third) handler argument is provided, it will be called when the event
+    is triggered; thus allowing this to be used as both a function and a
+    decorator.
+    """
+    # If "when" is called as a function, try to grab the handler from the
+    # arguments. If there's no handler, this must be a decorator based call.
+    handler = None
+    if args and (callable(args[0]) or is_awaitable(args[0])):
+        handler = args[0]
+    elif callable(kwargs.get("handler")) or is_awaitable(kwargs.get("handler")):
+        handler = kwargs.pop("handler")
+    # If the target is a string, it is the "older" use of `when` where it
+    # represents the name of a DOM event.
+    if isinstance(target, str):
+        # Extract the selector from the arguments or keyword arguments.
+        selector = args[0] if args else kwargs.pop("selector")
+        if not selector:
+            raise ValueError("No selector provided.")
+        # Grab the DOM elements to which the target event will be attached.
+        from pyscript.web import Element, ElementCollection
+
+        if isinstance(selector, str):
+            elements = document.querySelectorAll(selector)
+        elif isinstance(selector, Element):
+            elements = [selector._dom_element]
+        elif isinstance(selector, ElementCollection):
+            elements = [el._dom_element for el in selector]
+        else:
+            elements = selector if isinstance(selector, list) else [selector]
+
+    def decorator(func):
+        if "micropython" in sys.version.lower():
```

Comment:
didn't we introduce `pyscript.config["type"]` to disambiguate? 🤔 https://docs.pyscript.net/2024.10.2/api/#pyscriptconfig

Reply:
Good catch. Updating.